### PR TITLE
Assignment 2 - Code on R8, Detailed Test Plan, Code Reorganization

### PR DESCRIPTION
**Pull request by Chumingsun**

### CODE REVIEW

**Comment by SethDamiani** [comment]
Few quick notes:

- Your tests fail due to multiple test files sharing the same name, which pytest does not allow.
- The title of your 404 page is 'Profile', which should probably be changed to something like 'Page not found'
- For consistency, we should format our Detailed Test Plan in markdown instead of Word/PDF

After those are fixed I think we'll be good to merge.

**Review by SethDamiani — APPROVED** [review]

File: qa327/frontend.py [hunk 1]
```diff
@@ -28,7 +29,7 @@ def register_post():
     if password != password2:
         error_message = "The passwords do not match"
 
-    elif len(email) < 1:
+    elif (len(email) < 1 or not re.match(r"^[A-Za-z0-9\.\+_-]+@[A-Za-z0-9\._-]+\.[a-zA-Z]*$", email)):
```
Comment (atangent) [inline qa327/frontend.py]: This is a good line of code to validate the email format. 

However, I'm seeing this twice in this file. I think this line could be modularized into a function that returns `True` if the email is valid or `False` if the email is not valid. Something like
```
def is_valid_email(email):
  INVALID_EMAIL = (len(email) < 1 or \
      not re.match(r"^[A-Za-z0-9\.\+_-]+@[A-Za-z0-9\._-]+\.[a-zA-Z]*$", email))
  return False if INVALID_EMAIL else True
```
This ensures we follow DRY principles and makes the code cleaner too.

Then you could call it here like:
```
if not is_valid_email:
    # error message and/or re-render login page 
 ```
which you can do again the second time you need to check email format.

File: qa327/frontend.py [hunk 2]
```diff
@@ -57,7 +58,10 @@ def login_post():
     email = request.form.get('email')
     password = request.form.get('password')
     user = bn.login_user(email, password)
-    if user:
+
+    if (len(email) < 1 or not re.match(r"^[A-Za-z0-9\.\+_-]+@[A-Za-z0-9\._-]+\.[a-zA-Z]*$", email)):
```
Comment (atangent) [inline qa327/frontend.py]: This is where you repeat yourself

**Review by atangent — CHANGES_REQUESTED** [review]
I think parts of the code in `frontend.py` could be modularized

**Review by atangent — APPROVED** [review]
Looks good to me

**Pull request merged**